src/llmq/quorums_blockprocessor.cpp

```diff
@@ -0,0 +1,455 @@
+{
+    auto& params = Params().GetConsensus().llmqs.at(llmqType);
+
+    int nHeight = pindexPrev->nHeight + 1;
+    int quorumStartHeight = nHeight - (nHeight % params.dkgInterval);
+    if (quorumStartHeight >= pindexPrev->nHeight) {
+        return uint256();
+    }
+    auto quorumIndex = pindexPrev->GetAncestor(quorumStartHeight);
+    assert(quorumIndex);
+    return quorumIndex->GetBlockHash();
+}
+
+bool CQuorumBlockProcessor::HasMinedCommitment(Consensus::LLMQType llmqType, const uint256& quorumHash)
+{
+    auto key = std::make_pair(DB_MINED_COMMITMENT, std::make_pair((uint8_t)llmqType, quorumHash));
+    return evoDb.Exists(key);
+}
+
+bool CQuorumBlockProcessor::GetMinedCommitment(Consensus::LLMQType llmqType, const uint256& quorumHash, CFinalCommitment& ret)
+{
+    auto key = std::make_pair(DB_MINED_COMMITMENT, std::make_pair((uint8_t)llmqType, quorumHash));
+    return evoDb.Read(key, ret);
+}
+
+bool CQuorumBlockProcessor::HasMinableCommitment(const uint256& hash)
+{
+    LOCK(minableCommitmentsCs);
+    return minableCommitments.count(hash) != 0;
+}
+
+void CQuorumBlockProcessor::AddMinableCommitment(const CFinalCommitment& fqc)
+{
+    bool relay = false;
+    uint256 commitmentHash = ::SerializeHash(fqc);
+
+    {
+        LOCK(minableCommitmentsCs);
+
+        auto k = std::make_pair((Consensus::LLMQType) fqc.llmqType, fqc.quorumHash);
+        auto ins = minableCommitmentsByQuorum.emplace(k, commitmentHash);
+        if (ins.second) {
+            minableCommitments.emplace(commitmentHash, fqc);
+            relay = true;
+        } else {
+            auto& oldFqc = minableCommitments.at(ins.first->second);
+            if (fqc.CountSigners() > oldFqc.CountSigners()) {
+                // new commitment has more signers, so override the known one
+                ins.first->second = commitmentHash;
+                minableCommitments.erase(ins.first->second);
+                minableCommitments.emplace(commitmentHash, fqc);
+                relay = true;
+            }
+        }
+    }
+
+    // We only relay the new commitment if it's new or better then the old one
+    if (relay) {
+        CInv inv(MSG_QUORUM_FINAL_COMMITMENT, commitmentHash);
+        g_connman->RelayInv(inv, DMN_PROTO_VERSION);
+    }
+}
+
+bool CQuorumBlockProcessor::GetMinableCommitmentByHash(const uint256& commitmentHash, llmq::CFinalCommitment& ret)
+{
+    LOCK(minableCommitmentsCs);
+    auto it = minableCommitments.find(commitmentHash);
+    if (it == minableCommitments.end()) {
+        return false;
+    }
+    ret = it->second;
+    return true;
+}
+
+// Will return false if no commitment should be mined
+// Will return true and a null commitment if no minable commitment is known and none was mined yet
+bool CQuorumBlockProcessor::GetMinableCommitment(Consensus::LLMQType llmqType, const CBlockIndex* pindexPrev, CFinalCommitment& ret)
```

Comment:
Would probably make sense to extract all `false` checks into a separate function e.g. `IsMinableCommitmentRequired` or smth like that.

Reply:
Done